wakuv2/missing_messages.go

```diff
@@ -0,0 +1,272 @@
+	pubsubTopic   string
+	contentTopics []string
+	lastChecked   time.Time
+
+	ctx    context.Context
+	cancel context.CancelFunc
+}
+
+func (p TopicInterest) Equals(other TopicInterest) bool {
+	if p.peerID != other.peerID {
+		return false
+	}
+
+	if p.pubsubTopic != other.pubsubTopic {
+		return false
+	}
+
+	slices.Sort(p.contentTopics)
+	slices.Sort(other.contentTopics)
+
+	for i, contentTopic := range p.contentTopics {
+		if contentTopic != other.contentTopics[i] {
+			return false
+		}
+	}
+
+	return true
+}
+
+func (w *Waku) SetTopicsToVerifyForMissingMessages(peerID peer.ID, pubsubTopic string, contentTopics []string) {
+	w.topicInterestMu.Lock()
+	defer w.topicInterestMu.Unlock()
+
+	ctx, cancel := context.WithCancel(w.ctx)
+	newMissingMessageRequest := TopicInterest{
+		peerID:        peerID,
+		pubsubTopic:   pubsubTopic,
+		contentTopics: contentTopics,
+		lastChecked:   w.timesource.Now().Add(delay),
+		ctx:           ctx,
+		cancel:        cancel,
+	}
+
+	currMessageVerificationRequest, ok := w.topicInterest[pubsubTopic]
+	if !ok || !currMessageVerificationRequest.Equals(newMissingMessageRequest) {
+		if ok {
+			// If there is an ongoing request, we cancel it before replacing it
+			// by the new list. This can be probably optimized further by tracking
+			// the last time a content topic was synced, but might not be necessary
+			// since cancelling an ongoing request would mean cancelling just a single
+			// page of results
+			currMessageVerificationRequest.cancel()
+		}
+
+		w.topicInterest[pubsubTopic] = newMissingMessageRequest
+	}
+}
+
+func (w *Waku) checkForMissingMessages() {
+	defer w.wg.Done()
+	defer w.logger.Debug("checkForMissingMessages - done")
+
+	t := time.NewTicker(time.Minute)
+	defer t.Stop()
+
+	var semaphore = make(chan struct{}, 5)
+	for {
+		select {
+		case <-t.C:
+			w.logger.Debug("checking for missing messages...")
+			w.topicInterestMu.Lock()
+			for _, request := range w.topicInterest {
+				select {
+				case <-w.ctx.Done():
+					return
+				default:
+					semaphore <- struct{}{}
+					go func(r TopicInterest) {
+						w.FetchHistory(r)
+						<-semaphore
+					}(request)
+				}
+			}
+			w.topicInterestMu.Unlock()
+
+		case <-w.ctx.Done():
+			return
+		}
+	}
+}
+
+func (w *Waku) FetchHistory(missingHistoryRequest TopicInterest) {
+	for i := 0; i < len(missingHistoryRequest.contentTopics); i += maxContentTopicsPerRequest {
+		j := i + maxContentTopicsPerRequest
+		if j > len(missingHistoryRequest.contentTopics) {
+			j = len(missingHistoryRequest.contentTopics)
+		}
+
+		now := w.timesource.Now()
+		err := w.fetchMessagesBatch(missingHistoryRequest, i, j, now)
+		if err != nil {
+			if errors.Is(err, context.Canceled) {
+				return
+			} else {
+				w.logger.Error("could not fetch history", zap.Stringer("peerID", missingHistoryRequest.peerID), zap.String("pubsubTopic", missingHistoryRequest.pubsubTopic), zap.Strings("contentTopics", missingHistoryRequest.contentTopics))
+				continue
+			}
```

Comment:
`else` and `continue` are not really needed here I guess?

```suggestion
			if errors.Is(err, context.Canceled) {
				return
			}
		    w.logger.Error("could not fetch history", zap.Stringer("peerID", missingHistoryRequest.peerID), zap.String("pubsubTopic", missingHistoryRequest.pubsubTopic), zap.Strings("contentTopics", missingHistoryRequest.contentTopics))
```